src/pallene/uninitialized.lua

```diff
@@ -3,149 +3,102 @@
-
-    elseif tag == "ir.Cmd.Seq" then
-        for _, c in ipairs(cmd.cmds) do
-            local ft
-            ft, uninit = test(c, uninit, loop)
-            if not ft then return false end
+    local block_order = ir.get_depth_search_topological_sort(block_list)
+
+    local function block_analysis(block_i)
+        local block = block_list[block_i]
+        local uninit = uninit_sets[block_i]
+        local kill = kill_sets[block_i]
+        local changed = false
+        local jmp = ir.get_jmp(block)
+        local jmp_false = ir.get_jmpIfFalse(block)
+        if jmp  then
+            local c = merge_uninit(uninit_sets[jmp.target], uninit, kill)
+            changed = c or changed
         end
-        return true, uninit
-
-    elseif tag == "ir.Cmd.Break" then
-        assert(loop)
-        loop.is_infinite = false
-        merge(loop.uninit, uninit)
-        return false
-
-    elseif tag == "ir.Cmd.If" then
-        check_use(cmd.condition)
-
-        local ft1, uninit1 = test(cmd.then_, copy(uninit), loop)
-        local ft2, uninit2 = test(cmd.else_,      uninit , loop)
-
-        if ft1 and ft2 then
-            merge(uninit1, uninit2)
-            return true, uninit1
-        elseif ft1 then
-            return true, uninit1
-        elseif ft2 then
-            return true, uninit2
-        else
-            return false
-        end
-
-    elseif tag == "ir.Cmd.Loop" then
-        loop = new_loop()
-        test(cmd.body, uninit, loop)
-
-        if loop.is_infinite then
-            return false
-        else
-            return true, loop.uninit
-        end
-
-    elseif tag == "ir.Cmd.For" then
-        check_use(cmd.start)
-        check_use(cmd.limit)
-        check_use(cmd.step)
-
-        loop = new_loop()
-        loop.is_infinite = false
-        merge(loop.uninit, uninit)
-
-        uninit[cmd.dst] = nil
-        test(cmd.body, uninit, loop)
-
-        if loop.is_infinite then
-            return false
-        else
-            return true, loop.uninit
+        if jmp_false  then
+            local c = merge_uninit(uninit_sets[jmp_false.target], uninit, kill)
+            changed = c or changed
         end
+        return changed
+    end
 
-    elseif tagged_union.typename(cmd._tag) == "ir.Cmd" then
-        for _, val in ipairs(ir.get_srcs(cmd)) do
-            if val._tag == "ir.Value.LocalVar" then
-                -- `SetField` instructions can count as initializers when the target is an
-                -- upvalue box. This is because upvalue boxes are allocated, but not initialized
-                -- upon declaration.
-                if cmd._tag == "ir.Cmd.SetField" and cmd.rec_typ.is_upvalue_box then
-                    uninit[val.id] = nil
-                end
-                check_use(val.id)
-            end
+    repeat
+        local changed = false
+        for _,block_i in ipairs(block_order) do
+            changed = block_analysis(block_i) or changed
         end
+    until not changed
```

Comment:
Future work: consider a worklist approach. (we should discuss this next time we talk). Please open an issue so we don't forget.

Reply:
#609